cogs/snipe.py

```diff
@@ -0,0 +1,80 @@
+import discord
+from discord import Embed, File, Attachment, Message
+from discord.ext import commands
+from datetime import datetime, timedelta
+from collections import defaultdict
+
+import aiohttp, os
+import asyncio
+from io import BytesIO
+
+from typing import List, Tuple
+
+
+class DeletedMsg:
+    def __init__(self, msg: Message, attachments: List[Tuple[str, BytesIO]]):
+        self.msg = msg
+        self.attachments = attachments
+
+
+SNIPE_TIMER = 10
+
+
+class Snipe(commands.Cog):
+    def __init__(self, bot, http_client):
+        self.__bot = bot
+        self.http = http_client
+        self.deleted_messages = defaultdict(lambda: set())
+        self.__lock = asyncio.Lock()
+
+    async def cog_load(self):
+        await super().cog_load()
+        print("Snipe Cog loaded.")
+
+    async def save_attachment(self, attachment: Attachment):
+        async with self.http.get(attachment.url) as resp:
+            if resp.status != 200:
+                return attachment.filename, None
+            buf = BytesIO()
+            buf.write(await resp.read())
+            buf.seek(0)
+            return attachment.filename, buf
+
+    @commands.command(name="snipe")
+    async def snipe(self, ctx):
+        msgs = None
+        async with self.__lock:
+            msgs = self.deleted_messages[ctx.channel.id]
+
+        if not msgs:
+            await ctx.send(f"Nothing found")
+            return
+
+        for msg in msgs:
+            heading = Embed(description=msg.msg.content, color=0x00FF00)
+            heading.set_author(name=msg.msg.author.name)
+            embeds = [heading]
+            files = [File(buf, filename=fname) for fname, buf in msg.attachments if buf]
+            embeds.extend(
+                [Embed().set_image(url=f"attachment://{f.filename}") for f in files]
+            )
+            await ctx.send(embeds=embeds, files=files)
+
+    @commands.Cog.listener()
+    async def on_message_delete(self, message: discord.Message):
+        attachments = []
+        if message.attachments:
+            attachments = await asyncio.gather(
+                *[self.save_attachment(a) for a in message.attachments]
+            )
+        deleted_msg = DeletedMsg(message, attachments)
+        async with self.__lock:
+            self.deleted_messages[message.channel.id].add(deleted_msg)
+        await asyncio.sleep(SNIPE_TIMER)
```

Comment:
its not clear to me what happens if two messages are deleted in the same 10 second window. does it work as expected? only the last one should be retrieved

Reply:
This one is in a bit of a flux. Right now it returns all message deleted in the last x seconds. I was split on which one we want. Only the last message, or last message within x seconds. This is a simply change. Only need to add a clear when a new message arrives.

Reply:
ok thanks, that makes sense. was there a reason for sleep based deletion? the original method was based off timestamp which wouldn't need a sleep

Reply:
Arguably simpler code when returning all messages deleted in the last x seconds. There is no performance penalty of a sleeping coroutine.